ReSharper.FSharp/src/FSharp.Psi.Services/src/Generate/GenerateProvider.fs

```diff
@@ -198,28 +198,71 @@ type FSharpOverridingMembersBuilder() =
     inherit GeneratorBuilderBase<FSharpGeneratorContext>()
 
     let addNewLineIfNeeded (typeDecl: IFSharpTypeDeclaration) (typeRepr: ITypeRepresentation) =
-        if isNull typeRepr || typeDecl.StartLine <> typeRepr.StartLine then () else
+        if isNull typeRepr || typeDecl.Identifier.StartLine <> typeRepr.StartLine then () else
 
-        let currentIndent = typeRepr.Indent
-        let desiredIndent = typeDecl.Indent + typeDecl.GetIndentSize()
+        let indentSize = typeDecl.GetIndentSize()
+        let desiredIndent = typeDecl.Indent + indentSize
         
         addNodesBefore typeRepr.FirstChild [
             NewLine(typeRepr.GetLineEnding())
-            Whitespace(currentIndent)
+            Whitespace(desiredIndent)
         ] |> ignore
         
-        // to have good indents for begin/end keywords which might start out on different indents
-        shiftNode (-currentIndent) typeRepr
-        shiftNode desiredIndent typeRepr
+        let normalizeReprEnd (beginToken: ITokenNode) (endToken: ITokenNode) =
+            if beginToken.Indent > endToken.Indent then
+                let diff = beginToken.Indent - endToken.Indent
+                addNodeBefore endToken (Whitespace(diff))
         
+        match typeRepr with
+        | :? IClassRepresentation as classRepr ->
+            normalizeReprEnd classRepr.BeginKeyword classRepr.EndKeyword
+        | :? IStructRepresentation as structRepr ->
+            normalizeReprEnd structRepr.BeginKeyword structRepr.EndKeyword
+        | :? IRecordRepresentation as recordRepr ->
+            normalizeReprEnd recordRepr.LeftBrace recordRepr.RightBrace
+
+            if recordRepr.FieldDeclarations.Count > 1 then
+                recordRepr.FieldDeclarations
```

Comment:
Please extract `recordRepr.FieldDeclarations` to a binding, it's being calculated 4 times in this `if` expression.